src/recovery/store/in_mem.rs

```diff
@@ -331,58 +331,201 @@ fn write_discard_drops_key() {
 
 /// A progress store with all data in-memory.
 #[derive(Debug)]
-pub(crate) struct InMemProgress(HashMap<WorkerKey, BorderEpoch>);
+pub(crate) struct InMemProgress {
+    /// Stores the current execution.
+    ///
+    /// Defaults to `-1` so the initial call to [`resume_from`] will
+    /// generate the execution `0`. There will never be any recovery
+    /// progress data that will match execution `-1`, so as soon as we
+    /// start reading any, we'll bump to the correct execution.
+    ex: i128,
+    frontiers: HashMap<WorkerIndex, WorkerFrontier>,
+}
 
 impl InMemProgress {
-    pub(crate) fn new() -> Self {
-        Self(HashMap::new())
+    /// Init to the "default" progress of a cluster of a given
+    /// size.
+    ///
+    /// You have to specify a worker count so the initial data is
+    /// well-formed.
+    pub(crate) fn new(count: WorkerCount) -> Self {
+        let ex = Into::<i128>::into(<u64 as Timestamp>::minimum()) - 1;
+        let mut frontiers = HashMap::new();
+        for worker in count.iter() {
+            frontiers.insert(worker, WorkerFrontier(<u64 as Timestamp>::minimum()));
+        }
```

Comment:
This could be:

```rust
        let frontiers = count
            .iter()
            .map(|worker| (worker, WorkerFrontier(<u64 as Timestamp>::minimum())))
            .collect();
```

Also, since I see you do this more than once, it could be nice to have:

```rust
impl WorkerFrontier {
    pub fn minimum() -> Self {
        Self(<u64 as Timestamp>::minimum())
    }
}
```

So that you can write:
```rust
        let frontiers = count
            .iter()
            .map(|worker| (worker, WorkerFrontier::minimum()))
            .collect();
```